### DISCUSSION TITLE
Dynamic Configuration

### DISCUSSION THREAD

**Discussion by pyfisch** [archlinux/archinstall#150] [Ideas]
I want to script my Arch Linux installation process to ensure that I install the OS correctly, apply all my preferences and be able to quickly install it to a new machine.

I originally planned to write my own shell script but I discovered archinstall yesterday and I think it is a much better option for me. I want to fix some installation settings like keymap, locale and installed packages but be able to set some things like passwords and partitions during the actual installation.

AFAIK I have two options with archinstall right now: use [the guided install](https://github.com/archlinux/archinstall/blob/master/examples/guided.py) and pass some arguments via the command line, or I can write [my own installation script based on the minimal installation](https://github.com/archlinux/archinstall/blob/master/examples/minimal.py).  In the latter case I need to write my own input functions while in the former I can't customize the installation as much.

I want to write a script for my installation that looks somewhat like this:

```python3
# Create a new configuration, with some values fixed.
# Fields with the value PROMPT are specified during the installation.
config = Configuration(
	keymap="de",
	locale="de_DE.UTF-8",
	timezone="Europe/Berlin",
	encryption=Luks(password=PROMPT),
	# Overwrite "linux" from the default configuration.
	kernel="linux-zen",
	users=[
		User(name="pyfisch", password=PROMPT, sudo=True),
	],
	profile="gnome",
	packages=["firefox", "thunderbird"],
)
# Ask the user for values that aren't specified yet.
# Steps like keymap are skipped because the value is already set.
# Show prompt for packages so the user can add additional packages.
config.prompt(always=["packages"])
# Install the complete configuration without further user input.
config.install()

# Do some post-install steps like copying dotfiles.
```

Because the configuration and `prompt` don't change the system it can be tested without a virtual machine as a regular user.  The `prompt` method can have different front-ends e.g. a terminal dialog and a GUI.

The default guided installation would be a file that only contains:

```python3
from archinstall import Configuration

config = Configuration()
config.prompt()
config.install()
```

**Comment by Torxed** [comment]
I really do like the syntax of this.
We've been discussing declarative configurations for a long time but I never landed on something that looked clean and still supported user-input.. except a very brief experiment with [PYSON](https://github.com/Torxed/python-pyson).

I would be in favor of implementing this type of meta/wrapper around the installation process some how.
And I would also like if it supported taking the JSON output that the users see as a base/input (so we can re-run user reported issues) for faster debugging :)

**Reply by pyfisch** [reply]
> And I would also like if it supported taking the JSON output that the users see as a base/input (so we can re-run user reported issues) for faster debugging :)

I had in mind to store the fully specified (i.e. after prompt) configuration as JSON for debugging. This configuration can be loaded by other people to find the error cause.

**Reply by l4zy0n3** [reply]
I saw that archinstall spits out a json config at the end of input prompts, we could load the config from the same json structure

**Reply by Torxed** [reply]
I would very much like this. Bare in mind that any **key** that starts with `!` will be anonymized, such as `!password` or `!magic-key`. But when testing, we can manually change those before giving it to archinstall :)

**Reply by l4zy0n3** [reply]
Currently, install from config is working ( only format whole device), although I'm facing an issue, looks like it's unable to open fstab for some reason

**Reply by l4zy0n3** [reply]
logs at https://pastebin.com/puk6Tu9C

**Reply by Torxed** [reply]
This looks weird:
```
/usr/bin/pacstrap does not exist.
Could not strap in packages: 1
```
Not sure why those two pops up

**Reply by l4zy0n3** [reply]
> This looks weird:
> ```
> /usr/bin/pacstrap does not exist.
> Could not strap in packages: 1
> ```
> Not sure why those two pops up

Now I'm past this issue I was actually testing on a manjaro machine which doesn't have arch-install-scripts by default

Basic installation is working fine now ( only format whole device ) for now

**Reply by Torxed** [reply]
> Basic installation is working fine now ( only format whole device ) for now

That's totally acceptable.
The partitioner logic is getting a re-write for v2.2.0 (next release after v2.1.4), where we'll make it easier to distinguish between the formatting options :)

**Comment by Torxed** [comment]
I'll add to this discussion, that we haven't been able to conclude that very specific syntax now *(/yet)*.
But we do have more or less full support for JSON definitions now.
`--config` can swallow `/var/log/archinstall/user_configuration.json` files.
`--creds` can swallow `/var/log/archinstall/user_credentials.json` files.
And `--disk_layout` can take `/var/log/archinstall/user_disk_layout.json` files.

Code wise, `archinstall.Filesystem()` has a function called [load_layout()](https://github.com/archlinux/archinstall/blob/b7fcbf13bc1d3d1b1a6255d6eb2986149fa56982/archinstall/lib/disk/filesystem.py#L65) which also takes the above mentioned format.

After the release I'll get around to taking a look at making a config-reading-function for the general options that could be invoked along the lines of the suggestion possibly.

**Reply by hkaancaliskan** [reply]
--disk_layout not working but --disk_layouts working. I think we should add these arguments to readme.

**Reply by Torxed** [reply]
Apologies, spelling error. Yes they will be added, they were added not too many commits ago and are new for this upcoming release :)
